### PR TITLE
[do not merge] move startPage from divelist to main.

### PR DESCRIPTION
**Pull request by janiversen**
### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [ ] New feature
- [x ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

The startPage object does not belong in diveList, move to main

Make diveList a proper object of main, instead of indirect through the pageStack.

Smaller cleanups in main, startpage and divelist.

This PR sits on top of #2451, and #2451 should be merged first.

### CODE REVIEW

**Comment by dirkhh** [comment]
That seems to have a lot of additional commits in the PR.
The start page is part of the dive list because of the odd design that was chosen for the app... let me rephrase this - the people who worked on the initial version of the QML app are no longer part of the project - and a lot of the odd way the dive list is at the center of everything (and ALWAYS) the first page stems from that.
I'm ok with dis-entangling that - to a point.
There is "cleanup" and then there's "code thrash for cleanup" with very limited real world benefit.

I don't want to end up seeing large amounts of code moved around for no other reason than "it makes more sense this way". That's a slippery slope to introducing subtle bugs and causing me a ton of work.

**Comment by janiversen** [comment]
Yes there are all the commits from the PRs it depend on. This is a truly irritating "feature" of git itself, when you make a branch of a branch, e.g
master -> A -> B
B contains all the commits from A.
Now if you rebase A (due to review comment)
and then in B do "git rebase A", you often see a lot of conflict it the old commits from A.
I hope some day git will solve this problem, since that is one of my biggest time consumers.

I agree with your statement, except if you call these changes for "code thrash for cleanup", I do see a lot of development benefits in making the split, the old code had some interesting problems:
- loading StartPage in Qt Creator, to check the design did not work, because part of design (layout etc) was not in the class but in the object (instance in diveList).
- making changes to the divelist design often caused unexpected problems with startPage
- the reason for the awkward design seems to be an old limitation in PageStack (I assume), which are no longer the case.

I have tried only to make the smallest number of changes to make the split, and each commit focusing on a particular part of the move.

**Comment by dirkhh** [comment]
This one says "do not merge" in the description, not the title... so I'm holding off on reviewing

**Comment by janiversen** [comment]
my bad, forgot to edit the text.

**Comment by dirkhh** [comment]
I have rebased this locally with tiny changes. This seemed easiest instead of constantly sending you back to do more
Would you like to review the result, or should I just merge it?

**Comment by dirkhh** [comment]
my rebased version is here: https://github.com/Subsurface-divelog/subsurface/tree/newlogin3

**Comment by janiversen** [comment]
Feel free to do the merge. thanks.

**Comment by janiversen** [comment]
I believe you should be able to merge your version directly on top of master, and then do the set of reworked (setLogin*) PRs

**Comment by janiversen** [comment]
I just did an extra check, you have commits from other PR´s included:
from "mobile-widgets: make saveCredentials() an atom" and backwards are not part of this PR.

**Comment by dirkhh** [comment]
> I just did an extra check, you have commits from other PR´s included:
> from "mobile-widgets: make saveCredentials() an atom" and backwards are not part of this PR.

please scroll up on the web view of THIS PR (https://github.com/Subsurface-divelog/subsurface/pull/2452) and note that "mobile-widgets: make saveCredentials() an atom" is part of THIS PR. As are six more commits.

So if that and the one other, previously unmerged commit ("new mobile login: do not revert to old credentials") is part of the rebased version that I did, that shouldn't be surprising.

In another thread you said that you wanted me to trust you more. It's these things where you tell me I'm doing something wrong without (as far as I can tell) spending even 20 seconds on checking your own work that always make me worry. Whether it's the indentation of the `for` loop, or comments on type checking in connect calls, where you claim that reverting to the string based syntax is an improvement, when the Qt documentation quite explicitly states the opposite - https://doc.qt.io/qt-5/signalsandslots-syntaxes.html - I'll save you the time clicking on the link, because it appears you didn't click on the link and read the information the last time I commented on that:
> String-based connections type-check by comparing strings at run-time. There are three limitations with this approach:

> 1)    Connection errors can only be detected after the program has started running.
> 2)    Implicit conversions cannot be done between signals and slots.
> 3)    Typedefs and namespaces cannot be resolved.

>Limitations 2 and 3 exist because the string comparator does not have access to C++ type information, so it relies on exact string matching.

>In contrast, functor-based connections are checked by the compiler. The compiler catches errors at compile-time, enables implicit conversions between compatible types, and recognizes different names of the same type.

So, yeah, I have a hard time trusting the things you tell me in the comments to these PRs because of all these cases where I find that the things that you so confidently state aren't actually correct. And please note, I am still simply commenting about the things that you post and say here in these PR discussions. Things that you state as facts or strongly held opinion (like a commit must not touch multiple directories). I am NOT commenting about "you the developer" or "you the person".

Back to this PR.

a) if you have a smaller subset of commits in mind that should be merged, please close this PR, create a clean branch on top of master with just those commits, and create a new PR, referencing this one
b) if the commits that were in this PR (including the two not already merged ones that I have in the merged branch) are what you actually want to see merged into master, let me know and I'll merge my rebased branch

Thanks.

**Comment by janiversen** [comment]
Just a note the extra commits, are NOT because I added them, it is something git do:

you do 
git checkout -b A
git commit X1
git commit X2
git checkout -b B
git commit X3
git push .... 
And you see nicely that the PR contains X1, X2, X3 
Which is why I am struggling every time I have a series of branches, and do a rebase in the one of them.

I will resubmit this one.

**Comment by janiversen** [comment]
oooh btw, you do not need to manually remove the commits, I do a 
git rebase subsurface-master
after each PR you merge/close.

**Comment by dirkhh** [comment]
the duplicate commits are caused by the way you create those PRs. I understand that.
Trust me. I do understand how git works. I will bet you any sum you want that I have used git longer than you have.

Instead of having all these arguments, I will simply not merge things on GitHub but merge them locally and push like I did with the last couple.

I am optimizing for my time here. Selfish, arrogant, arbitrary. That's me.

**Comment by janiversen** [comment]
Yes for sure you know git a lot better, I used SVN for a very loong time, so I am novice in git land....and an alternative solution was to tell me, what I do wrong, so I can avoid causing us both work.

**Comment by janiversen** [comment]
rebased and cleaned to only contain the commit needed.

Will now rebase all planner PRs

**Comment by dirkhh** [comment]
I am ok with the way you do the PRs, now that I understand what you do and how you get to where you are.
Dealing with this on my end is simple by doing the merging on the command line instead of on GitHub. No worries.

File: mobile-widgets/qml/DiveList.qml
Comment (dirkhh) [reply mobile-widgets/qml/DiveList.qml:415]: ah. it's in the next commit.
that seems unfortunate, but whatever.

File: mobile-widgets/qml/DiveList.qml
Comment (dirkhh) [inline mobile-widgets/qml/DiveList.qml:46]: useful during debugging, maybe not something we should keep around.
and yes, I am the absolute worst case offender in keeping them around...
BTW - I have switched to using the `appendTextToLog` function as that at least adds time stamps to the log which makes things easier to read.

File: mobile-widgets/qml/DiveList.qml
Comment (dirkhh) [inline mobile-widgets/qml/DiveList.qml]: I've read this five times and it's still not clear to me if this is functionally equivalent or not.
Which is mostly a sign that the existing code (mostly done by me) is bat-shit-crazy.
So I'll trust you that you've thought this through and will test the end result :-)

**Review by dirkhh — APPROVED** [review]
I think this is reasonable. Some comments, but I think what I want to do is play with this some more locally to make sure I truly understand how things changed. That's why I marked it as Approved. Unless I come back here with more comments, I don't think you need to do anything here.

**Comment by dirkhh** [comment]
ARG. So trying to run this on the desktop on my Mac I get 
```
INFO: QQmlApplicationEngine failed to load component
INFO: qrc:/qml/main.qml:820 Type Export unavailable
qrc:/qml/Export.qml:15 Type FileDialog unavailable
file:///.../subsurface/build-mobile/Subsurface-mobile.app/Contents/Resources/qml/QtQuick/Dialogs/DefaultFileDialog.qml:47 module "Qt.labs.folderlistmodel" is not installed
file:///.../subsurface/build-mobile/Subsurface-mobile.app/Contents/Resources/qml/QtQuick/Dialogs/DefaultFileDialog.qml:48 module "Qt.labs.settings" is not installed
file:///.../subsurface/build-mobile/Subsurface-mobile.app/Contents/Resources/qml/QtQuick/Dialogs/DefaultFileDialog.qml:47 module "Qt.labs.folderlistmodel" is not installed
file:///.../subsurface/build-mobile/Subsurface-mobile.app/Contents/Resources/qml/QtQuick/Dialogs/DefaultFileDialog.qml:48 module "Qt.labs.settings" is not installed
```

almost certainly a sign that I am missing some optional Qt components... but also a sign that we added dependencies that I wasn't aware we had added...

The only reason I'm posting this here is that this will slow me down until I have that working again (yes, I could go back to my laptop and test there, but I much rather do it on the crazy fast iMac)

**Comment by dirkhh** [comment]
Using my laptop, tethered via my phone. Building this branch on Mac with Qt 5.14.0
I get the weirdest results.
I enter incorrect credentials.
On the next screen, the email input is replaced by stars, the password is shown in clear text
Whatever I try I cannot get it to send the correct email/password to the server
But I also can't quite make sense of the information I get in the debug log
I restart, I enter correct credentials, everything works as planned.
This needs more testing / tracing, but there's definitely something not quite right, yet.

**Comment by janiversen** [comment]
Sounds a like you are not using 5.13, at least I saw a similar qt.labs missing in 5.11

**Comment by janiversen** [comment]
ups see now you are using 5.14 I have not tested that. Might be we have to cp the foldersettings module.

**Comment by janiversen** [comment]
with all this rebasing back and forth I might have introduced an error. Let me do the full test tomorrow instead of using your sparse time

**Comment by dirkhh** [comment]
Power is back.
That error was on the iMac which was on 5.13.0.
I didn't see the error on 5.14 on the laptop (and I'll upgrade the iMac to 5.14 as well)

**Comment by dirkhh** [comment]
the test with the strange results was done on the laptop with 5.14. Once I upgraded the iMac and have completed a clean full build on that, I'll do more testing of this branch there as well.

**Comment by dirkhh** [comment]
OK, I was able to reproduce what I had happen.
Enter a new email address and password.
The screen changes to this which I found confusing: what's the blacked out text under cloud credentials? If I edit that, do I change the password? and then the PIN - but since the layout is EXACTLY as the email/password setup page, I didn't even realize that this now asked me for the PIN.
Screen Shot 2019-12-24 at 2 42 51 PM

**Comment by janiversen** [comment]
hmmm that is CloudCredentials.qml. If that logic is changed, then there is a bigger problem. I will investigate tomorrow

**Comment by dirkhh** [comment]
I did the thing I should have done right away, going back to master and of course this isn't caused by this PR. Instead this is caused by forgetting to change one instance of `rootItem.showPin` to `prefs.showPin` in a4d299e01e

I'll fix that and continue testing / merge this once I'm happy.

**Comment by dirkhh** [comment]
merged

**Pull request closed without merging**
